locator/tablets.cc

```diff
@@ -434,6 +434,58 @@ load_stats& load_stats::operator+=(const load_stats& s) {
     return *this;
 }
 
+tablet_range_splitter::tablet_range_splitter(schema_ptr schema, const tablet_map& tablets, host_id host, const dht::partition_range_vector& ranges)
+    : _schema(std::move(schema))
+    , _ranges(ranges)
+    , _ranges_it(_ranges.begin())
+{
+    // Filter all tablets and save only those that have a replica on the specified host.
+    for (auto tid = std::optional(tablets.first_tablet()); tid; tid = tablets.next_tablet(*tid)) {
+        const auto& tablet_info = tablets.get_tablet_info(*tid);
+
+        auto replica_it = std::ranges::find_if(tablet_info.replicas, [&] (auto&& r) { return r.host == host; });
+        if (replica_it == tablet_info.replicas.end()) {
+            continue;
+        }
+
+        _tablet_ranges.emplace_back(range_split_result{replica_it->shard, dht::to_partition_range(tablets.get_token_range(*tid))});
+    }
+    _tablet_ranges_it = _tablet_ranges.begin();
+}
+
+std::optional<tablet_range_splitter::range_split_result> tablet_range_splitter::operator()() {
+    if (_ranges_it == _ranges.end() || _tablet_ranges_it == _tablet_ranges.end()) {
+        return {};
+    }
+
+    dht::ring_position_comparator cmp(*_schema);
+
+    while (_ranges_it != _ranges.end()) {
+        // First, skip all tablet-ranges that are completely before the current range.
+        while (_ranges_it->before(_tablet_ranges_it->range.end()->value(), cmp)) {
+            ++_tablet_ranges_it;
```

Comment:
The call to ->value() ignores information about inclusiveness/exclusiveness. Should use range methods.

Reply:
Yes, this might lead us to think there is an intersection, but when we call `intersect()` we do take inclusiveness/exclusiveness into consideration and we will produce an empty intersection.

Reply:
In other words this is not a correctness issue, but I will fix it.

Reply:
There is no range version, I will add one.

Reply:
Added range version in v5 and used it instead of the point version.